crypto/bn/asm/rsaz-2k-avxifma.pl

```diff
@@ -0,0 +1,863 @@
+my $mask52     = "%rax";
+my $acc0_0     = "%r9";
+my $acc0_0_low = "%r9d";
+my $acc0_1     = "%r15";
+my $acc0_1_low = "%r15d";
+my $b_ptr      = "%r11";
+
+my $iter = "%ebx";
+
+my $zero = "%ymm0";
+my $Bi   = "%ymm1";
+my $Yi   = "%ymm2";
+my $Yi_xmm   = "%xmm2";
+my ($R0_0,$R0_0h,$R1_0,$R1_0h,$R2_0) = ("%ymm3",map("%ymm$_",(5..8)));
+my ($R0_1,$R0_1h,$R1_1,$R1_1h,$R2_1) = ("%ymm4",map("%ymm$_",(9..12)));
+
+# Registers mapping for normalization.
+my ($T0,$T0h,$T1,$T1h,$T2,$tmp) = ("$zero", "$Bi", "$Yi", map("%ymm$_", (13..15)));
+my $T0_xmm   = "%xmm0";
+
+sub amm52x20_x1() {
+# _data_offset - offset in the |a| or |m| arrays pointing to the beginning
+#                of data for corresponding AMM operation;
+# _b_offset    - offset in the |b| array pointing to the next qword digit;
+my ($_data_offset,$_b_offset,$_acc,$_R0,$_R0h,$_R1,$_R1h,$_R2,$_k0) = @_;
+$code.=<<___;
+    movq    $_b_offset($b_ptr), %r13             # b[i]
+
+    vpbroadcastq    $_b_offset($b_ptr), $Bi      # broadcast b[i]
+    movq    $_data_offset($a), %rdx
+    mulx    %r13, %r13, %r12                     # a[0]*b[i] = (t0,t2)
+    addq    %r13, $_acc                          # acc += t0
+    movq    %r12, %r10
+    adcq    \$0, %r10                            # t2 += CF
+
+    movq    $_k0, %r13
+    imulq   $_acc, %r13                          # acc * k0
+    andq    $mask52, %r13                        # yi = (acc * k0) & mask52
+
+    vmovq   %r13, $Yi_xmm
+    vpbroadcastq    $Yi_xmm, $Yi                 # broadcast y[i]
+    movq    $_data_offset($m), %rdx
+    mulx    %r13, %r13, %r12                     # yi * m[0] = (t0,t1)
+    addq    %r13, $_acc                          # acc += t0
+    adcq    %r12, %r10                           # t2 += (t1 + CF)
+
+    shrq    \$52, $_acc
+    salq    \$12, %r10
+    or      %r10, $_acc                          # acc = ((acc >> 52) | (t2 << 12))
+
+    lea -168(%rsp), %rsp
+    {vex} vpmadd52luq `$_data_offset+64*0`($a), $Bi, $_R0
+    {vex} vpmadd52luq `$_data_offset+64*0+32`($a), $Bi, $_R0h
+    {vex} vpmadd52luq `$_data_offset+64*1`($a), $Bi, $_R1
+    {vex} vpmadd52luq `$_data_offset+64*1+32`($a), $Bi, $_R1h
+    {vex} vpmadd52luq `$_data_offset+64*2`($a), $Bi, $_R2
+
+    {vex} vpmadd52luq `$_data_offset+64*0`($m), $Yi, $_R0
+    {vex} vpmadd52luq `$_data_offset+64*0+32`($m), $Yi, $_R0h
+    {vex} vpmadd52luq `$_data_offset+64*1`($m), $Yi, $_R1
+    {vex} vpmadd52luq `$_data_offset+64*1+32`($m), $Yi, $_R1h
+    {vex} vpmadd52luq `$_data_offset+64*2`($m), $Yi, $_R2
+
+    # Shift accumulators right by 1 qword, zero extending the highest one
+    vmovdqu $_R0,   `32*0`(%rsp)
+    vmovdqu $_R0h,  `32*1`(%rsp)
+    vmovdqu $_R1,   `32*2`(%rsp)
+    vmovdqu $_R1h,  `32*3`(%rsp)
+    vmovdqu $_R2,   `32*4`(%rsp)
+    movq \$0, `32*5`(%rsp)
+
+    vmovdqu `32*0 + 8`(%rsp), $_R0
+    vmovdqu `32*1 + 8`(%rsp), $_R0h
+    vmovdqu `32*2 + 8`(%rsp), $_R1
+    vmovdqu `32*3 + 8`(%rsp), $_R1h
+    vmovdqu `32*4 + 8`(%rsp), $_R2
```

Comment:
This fragment of code may result in store to load penalty but I haven't found a better option here. `alignr` operation on `ymm` registers is unusable here.

Reply:
It's important to note that the AVX512 `valignq` instruction is not supported on Sierra Forest processors, and using the `alignr` instruction is not a viable option here. We are experimenting with our initial attempt to emulate the `valignq` instruction using the `vpermq` instruction, which rearranges the Qword elements within a `ymm` register according to a fixed mask, as demonstrated in the following code snippet.
![image](https://github.com/user-attachments/assets/2cfee6e7-a366-43d8-8b1b-729ff2ce7cd8)

While the `vpermq`-based approach does have the benefit of not requiring any memory loads or stores, we have discovered that a memory-based implementation for the shift operation, as introduced in this pull request, could potentially enhance performance by **6%-30%** in speed benchmarks. Given that, it might be beneficial to retain the memory-based implementation.

Reply:
Thanks @ZhiguoZh. Agree with you. I also prototyped an alternative option and it was producing worse performance than the applied above memory based approach. No change needed here - just a note to take for others reading this code.